### PR TITLE
[glsl-out] Fix shader translation errors

### PR DESCRIPTION
**Pull request by Gordon-F**
Possibly fix for #1115 and #1085.

@dhardy Could you retest your issue with this commits?

### CODE REVIEW

File: src/back/glsl/mod.rs [hunk 1]
```diff
@@ -2329,6 +2348,54 @@ impl<'a, W: Write> Writer<'a, W> {
...
+    /// Helper function that write string with default zero initialization for supported types
+    fn write_zero_init_value(&mut self, inner: &TypeInner) -> BackendResult {
+        match *inner {
+            TypeInner::Scalar { kind, .. } => {
+                self.write_zero_init_scalar(kind)?;
+            }
+            TypeInner::Vector { size, kind, .. } => {
+                self.write_value_type(inner)?;
+                write!(self.out, "(")?;
+                for _ in 0..(size as usize) - 1 {
```
Comment (kvark) [inline src/back/glsl/mod.rs:2360]: easier to write this as `for _ in 1 .. size`

File: src/back/glsl/mod.rs [hunk 2]
```diff
@@ -2329,6 +2348,54 @@ impl<'a, W: Write> Writer<'a, W> {
...
+                }
+                // write last parameter without comma and space
+                self.write_zero_init_scalar(kind)?;
+                write!(self.out, ")")?;
+            }
+            TypeInner::Matrix { columns, rows, .. } => {
+                let number_of_components = (columns as usize) * (rows as usize);
+                self.write_value_type(inner)?;
+                write!(self.out, "(")?;
+                for _ in 0..number_of_components - 1 {
```
Comment (kvark) [inline src/back/glsl/mod.rs:2372]: nit: same here

**Review by kvark — APPROVED** [review]

**Review by kvark — APPROVED** [review]

**Comment by dhardy** [comment]
@Gordon-F took a while to discover that I needed `resolver = 2` to get anywhere... then I needed to patch a work of `wgpu` to use the latest `naga`.

After that, yes, it appears to fix both issues mentioned. The initial draw of my example apps is perfect, but then graphical corruption occurs in redraws (depending on the app).
gallery-corrupt

**Comment by Gordon-F** [comment]
@Gordon-F Thank you!

**Comment by dhardy** [comment]
I closed the first; still need to debug that corruption however.

**Pull request merged**